ml4gw/transforms/qtransform.py

```diff
@@ -0,0 +1,444 @@
+        For each frequency, a `QTile` is created.
+        """
+        minf, maxf = self.frange
+        fcum_mismatch = (
+            math.log(maxf / minf) * (2 + self.q**2) ** (1 / 2.0) / 2.0
+        )
+        deltam = 2 * (self.mismatch / 3.0) ** (1 / 2.0)
+        nfreq = int(max(1, math.ceil(fcum_mismatch / deltam)))
+        fstep = fcum_mismatch / nfreq
+        fstepmin = 1 / self.duration
+
+        freq_base = math.exp(2 / ((2 + self.q**2) ** (1 / 2.0)) * fstep)
+        freqs = torch.Tensor([freq_base ** (i + 0.5) for i in range(nfreq)])
+        freqs = (minf * freqs // fstepmin) * fstepmin
+        return torch.unique(freqs)
+
+    def get_max_energy(
+        self, fsearch_range: List[float] = None, dimension: str = "both"
+    ):
+        """
+        Gets the maximum energy value among the QTiles. The maximum can
+        be computed across all batches and channels, across all channels,
+        across all batches, or individually for each channel/batch
+        combination. This could be useful for allowing the use of different
+        Q values for different channels and batches, but the slicing would
+        be slow, so this isn't used yet.
+
+        Optionally, a pair of frequency values can be specified for
+        `fsearch_range` to restrict the frequencies in which the maximum
+        energy value is sought.
+        """
+        allowed_dimensions = ["both", "neither", "channel", "batch"]
+        if dimension not in allowed_dimensions:
+            raise ValueError(f"Dimension must be one of {allowed_dimensions}")
+
+        if self.qtiles is None:
+            raise RuntimeError(
+                "Q-tiles must first be computed with .compute_qtiles()"
+            )
+
+        if fsearch_range is not None:
+            start = min(torch.argwhere(self.freqs > fsearch_range[0]))
+            stop = min(torch.argwhere(self.freqs > fsearch_range[1]))
+            qtiles = self.qtiles[start:stop]
+        else:
+            qtiles = self.qtiles
+
+        if dimension == "both":
+            return max([torch.max(qtile) for qtile in qtiles])
+
+        max_across_t = [torch.max(qtile, dim=-1).values for qtile in qtiles]
+        max_across_t = torch.stack(max_across_t, dim=-1)
+        max_across_ft = torch.max(max_across_t, dim=-1).values
+
+        if dimension == "neither":
+            return max_across_ft
+        if dimension == "channel":
+            return torch.max(max_across_ft, dim=0).values
+        if dimension == "batch":
+            return torch.max(max_across_ft, dim=-1).values
+
+    def compute_qtiles(self, X: torch.Tensor, norm: str = "median"):
+        """
+        Take the FFT of the input timeseries and calculate the transform
+        for each `QTile`
+        """
+        # Computing the FFT with the same normalization and scaling as GWpy
+        X = torch.fft.rfft(X, norm="forward")
+        X[..., 1:] *= 2
+        self.qtiles = [qtile(X, norm) for qtile in self.qtile_transforms]
```

Comment:
Is there a reason for updating this attribute in place as oppposed to returning the qtiles, and letting downstream functions handle how to process it?

Mostly just a readability / code-clarity thing

Reply:
It's not updating in place, `self.qtile_transforms` is the list of transformations and `self.qtiles` is the list of transformed data (though, I see what you mean about readability with those two names). I think my thought process had been that the transformed data should live with the Q-transform object once it had been computed, so I wanted attributes for both the transforms and the transformed data.

I could have this function return `self.qtiles` in addition to assigning the attribute.